tokio/tests/task_local.rs

```diff
@@ -117,3 +117,209 @@ async fn task_local_available_on_completion_drop() {
+            Poll::Ready(())
+        }
+    }
+    impl Drop for MyFuture {
+        fn drop(&mut self) {
+            println!("task id: {}", task::id());
+        }
+    }
+
+    tokio::spawn(MyFuture).await.unwrap();
+}
+
+#[cfg(tokio_unstable)]
+#[tokio::test(flavor = "multi_thread")]
+async fn task_id_future_destructor_abort() {
+    use tokio::task;
+
+    struct MyFuture;
+    impl Future for MyFuture {
+        type Output = ();
+
+        fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<()> {
+            Poll::Pending
+        }
+    }
+    impl Drop for MyFuture {
+        fn drop(&mut self) {
+            println!("task id: {}", task::id());
+        }
+    }
+
+    tokio::spawn(MyFuture).abort();
+}
+
+#[cfg(tokio_unstable)]
+#[tokio::test(flavor = "current_thread")]
+async fn task_id_output_destructor_handle_dropped_before_completion() {
+    use tokio::task;
+
+    struct MyOutput;
+    impl Drop for MyOutput {
+        fn drop(&mut self) {
+            println!("task id: {}", task::id());
+        }
+    }
+
+    struct MyFuture {
+        tx: Option<oneshot::Sender<()>>,
+    }
+    impl Future for MyFuture {
+        type Output = MyOutput;
+
+        fn poll(mut self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+            let _ = self.tx.take().unwrap().send(());
+            Poll::Ready(MyOutput)
+        }
+    }
+    impl Drop for MyFuture {
+        fn drop(&mut self) {
+            println!("task id: {}", task::id());
+        }
+    }
+
+    let (tx, rx) = oneshot::channel();
+    let handle = tokio::spawn(MyFuture { tx: Some(tx) });
+    drop(handle);
+    rx.await.unwrap();
+}
+
+#[cfg(tokio_unstable)]
+#[tokio::test(flavor = "current_thread")]
+async fn task_id_output_destructor_handle_dropped_after_completion() {
+    use tokio::task;
+
+    struct MyOutput;
+    impl Drop for MyOutput {
+        fn drop(&mut self) {
+            println!("task id: {}", task::id());
+        }
+    }
+
+    struct MyFuture {
+        tx: Option<oneshot::Sender<()>>,
+    }
+    impl Future for MyFuture {
+        type Output = MyOutput;
+
+        fn poll(mut self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+            let _ = self.tx.take().unwrap().send(());
+            Poll::Ready(MyOutput)
+        }
+    }
+    impl Drop for MyFuture {
+        fn drop(&mut self) {
+            println!("task id: {}", task::id());
+        }
+    }
+
+    let (tx, rx) = oneshot::channel();
+    let handle = tokio::spawn(MyFuture { tx: Some(tx) });
+    rx.await.unwrap();
+    drop(handle);
+}
+
+#[cfg(tokio_unstable)]
+#[test]
+fn task_try_id_outside_task() {
+    use tokio::task;
+
+    assert_eq!(None, task::try_id());
+}
+
+#[cfg(tokio_unstable)]
+#[test]
+fn task_try_id_inside_block_on() {
+    use tokio::runtime::Runtime;
+    use tokio::task;
+
+    let rt = Runtime::new().unwrap();
+    rt.block_on(async {
+        assert_eq!(None, task::try_id());
+    });
+}
```

Comment:
i'm not sure if the `task_local` file here is the correct place for the task ID tests --- this module has the tests for the `task_local!` macro, which isn't really related to the task ID APIs.

i think we should probably add a new `task_id.rs` file in `tokio/tests` for these tests, instead.